Content/Items/Misc/SoilgunFiles/Weapons.Soilgun.cs

```diff
@@ -81,464 +89,338 @@ public override void AddRecipes()
 
-					switch (item.type)
-					{
-						case ItemID.SandBlock: ammoTooltip = new TooltipLine(Mod, "AmmoInfoTooltip", "When used with the Soilgun or Earthduster, it will fire out blocks of sand that split into many grains of sand upon death"); break;
-						case ItemID.CrimsandBlock: ammoTooltip = new TooltipLine(Mod, "AmmoInfoTooltip", "When used with the Soilgun or Earthduster, it will fire out blocks of Crimsand that steal life from hit enemies"); break;
-						case ItemID.EbonsandBlock: ammoTooltip = new TooltipLine(Mod, "AmmoInfoTooltip", "When used with the Soilgun or Earthduster, it will fire out blocks of Ebonsand that apply stacks of Haunted to enemies"); break;
-						case ItemID.PearlsandBlock: ammoTooltip = new TooltipLine(Mod, "AmmoInfoTooltip", "When used with the Soilgun or Earthduster, it will fire out blocks of Pearlsand that home in on enemies"); break;
-						case ItemID.DirtBlock: ammoTooltip = new TooltipLine(Mod, "AmmoInfoTooltip", "When used with the Soilgun or Earthduster, it will fire out blocks of dirt"); break;
-						case ItemID.SiltBlock: ammoTooltip = new TooltipLine(Mod, "AmmoInfoTooltip", "When used with the Soilgun or Earthduster, it will fire out blocks of silt, that spawn coins upon hitting enemies"); break;
-						case ItemID.SlushBlock: ammoTooltip = new TooltipLine(Mod, "AmmoInfoTooltip", "When used with the Soilgun or Earthduster, it will fire out blocks of slush that cause hit enemies to have icicles impale them\nHitting and enemy with more than 10 icicles causes all icicles to shatter, causing large amounts of damage"); break;
-						case ItemID.MudBlock: ammoTooltip = new TooltipLine(Mod, "AmmoInfoTooltip", "When used with the Soilgun or Earthduster, it will fire out blocks of mud that bounce off tiles and enemies"); break;
-					}
+			if (!CanHold && !Shot)
+			{
+				if (ChargeProgress >= 0.35f)
+				{
+					Shoot();
 
-					tooltips.Add(ammoTooltip);
-					ammoTooltip.OverrideColor = new Color(202, 148, 115);
-				}
+					Projectile.timeLeft = 45;
+					updateVelocity = false;
+					Shot = true;
+				}				
 			}
-		}
-	}
-
-	//TODO: Port to stackable buffs
-	class SoilgunGlobalNPC : GlobalNPC
-	{
-		public const int MAXHAUNTEDSTACKS = 20;
 
-		public const int MAXVITRICSHARDS = 10;
+			if (Charge == 0f)
+			{
+				ghostProjectile.SetDefaults(projectileID);
+				MaxCharge = CombinedHooks.TotalUseTime(Owner.HeldItem.useTime, Owner, Owner.HeldItem);
+				Projectile.velocity = Owner.DirectionTo(Main.MouseWorld);
+			}
 
-		public int GlassAmount;
+			UpdateHeldProjectile(!Shot);
 
-		public int GlassPlayerID;
+			if (Charge < MaxCharge && !Shot)
+			{
+				Charge++;
+			}
+			
+			if (Charge >= MaxCharge && !flashed)
+			{
+				flashTimer = 25;
+				flashed = true;
+			}	
+		}
 
-		public int HauntedStacks;
+		public override bool PreDraw(ref Color lightColor)
+		{
+			if (Charge <= 2)
+				return false;
 
-		public int HauntedTimer;
+			Texture2D tex = ModContent.Request<Texture2D>(Texture).Value;
```

Comment:
Needs to switch to AssGen like all the others, in general you should never need to call ModContent.Request for textures now